pkg/controller.v2/clustertrainingruntime_controller.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllerv2
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	kubeflowv2 "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v2alpha1"
+)
+
+type ClusterTrainingRuntimeReconciler struct {
+	log      logr.Logger
+	client   client.Client
+	recorder record.EventRecorder
+}
+
+func NewClusterTrainingRuntimeReconciler(client client.Client, recorder record.EventRecorder) *ClusterTrainingRuntimeReconciler {
+	return &ClusterTrainingRuntimeReconciler{
+		log:      ctrl.Log.WithName("clustertrainingruntime-reconciler"),
+		client:   client,
+		recorder: recorder,
+	}
+}
+
+func (r *ClusterTrainingRuntimeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
Do we need to have reconcilers for Training Runtimes since we don't do any orchestration for them initially ?

Reply:
We need this reconciler since the TrainingRuntime reconcilers have only event handlers and predicators.

Reply:
I will implement event handlers and predicators in the third phase.

1st phase: this PR.
2nd phase: JobSet builder interface.
3rd phase: actual reconciler and event handlers.

Reply:
What kind of event handlers and predicators are you planning to implement for Training Runtime ?

Reply:
Sure. I'm planning the following handlers, but I may add more event handling mechanisms in the actual implementation steps.

Create: Trigger the TrainJob reconcilers by triggering the generic event of the TrainJob reconcilers.
Update: Ignore the Update event since the webhook can be disabled by webhookConfiguration failurePolicy.
Delete: Ignore the Delete event when the Runtime is still used by any TrainJob since the webhook can be disabled by webhookConfiguration failurePolicy.
Generic: None

Reply:
@tenzen-y For the Create event do you mean: Trigger the `TrainingRuntime` reconcilers by triggering the generic event of the `TrainingRuntime` reconcilers ?
For the Delete, do you mean that user can't delete TrainingRuntime if it used by any TrainJob ?

Reply:
> For the Create event do you mean: Trigger the TrainingRuntime reconcilers by triggering the generic event of the TrainingRuntime reconcilers ?

No, we do not trigger the TrainingRuntime reconciler; we just trigger the TrainJob reconciler by the TrainingRuntime event handler.

> For the Delete, do you mean that user can't delete TrainingRuntime if it used by any TrainJob ?

Yes, that's right. We should prevent accidently deleting the TrainingRuntime when the TrainingRuntime is referred by any TranJob.

Reply:
>No, we do not trigger the TrainingRuntime reconciler; we just trigger the TrainJob reconciler by the TrainingRuntime event handler.

@tenzen-y What orchestration for the TrainJob are we going to do when we trigger such event ?

Reply:
Let me show the actual step in the following:

0. During setting up controllers, we pass the go channel for TrainJob Generic EventHandler to the TrainigRuntime controller.
1. TrainingRuntime Controller: Get TrainingRuntime creation event
2. TrainingRuntime Controller: Send the object to the TrainJob controller using go channel
3. TrainJob Controller: Start reconciling

This is a generic technique to start reconciling by out-of-reconcilers. I will implement the mechanism in the 3rd step.

Reply:
I synced with @andreyvelich offline.
I was supposed to provide functionality creation of TraininigRuntime after the TrainJob was created.
So, I added TrainingRuntime reconcilers. But, we decided not to allow user to create TrainJob before the TrainingRuntime is created in the first iteration. So, users need to create the TrainigRuntime before the TrainJob is created. 
Additionally, we decided to handle the deletion of TrainingRuntim by finalizers, and then we never used to reconciler for deletion event for TrainingRuntime.